File: benchmarks/benchmark_oecluster_fingerprint.py
#!/usr/bin/env python3
"""Compare OEFP dense fingerprint kernels against oecluster fingerprint paths."""

from __future__ import annotations

import argparse
import os
import statistics
import sys
import time
from collections.abc import Callable, Sequence
from pathlib import Path
from typing import Any

import numpy as np
from openeye import oechem, oegraphsim


_REPO_ROOT = Path(__file__).resolve().parents[1]
_DEFAULT_OECLUSTER_SOURCE_DIR = Path(
    os.environ.get("OEFP_OECLUSTER_SOURCE_DIR", "/Users/johnss51/Development/cpp/oecluster")
)


def _prepend_source_paths(oecluster_source_dir: Path) -> None:
    """Prefer local source-tree packages when the benchmark runs from checkout."""
    sys.path.insert(0, str(_REPO_ROOT / "python"))
    oecluster_python = oecluster_source_dir / "python"
    if oecluster_python.is_dir():
        sys.path.insert(0, str(oecluster_python))


def _seed_smiles() -> list[str]:
    return [
        "c1ccccc1",
        "c1ccc(O)cc1",
        "c1ccncc1",
        "CCCCCCCC",
        "CCO",
        "CCN",
        "CC(=O)O",
        "c1ccc(Cl)cc1",
    ]


def build_molecules(count: int) -> list[Any]:
    """Build a deterministic molecule set for repeatable benchmark inputs."""
    seeds = _seed_smiles()
    mols = []
    for index in range(count):
        mol = oechem.OEGraphMol()
        if not oechem.OESmilesToMol(mol, seeds[index % len(seeds)]):
            raise RuntimeError("OESmilesToMol failed")
        mols.append(mol)
    return mols


def time_call(fn: Callable[[], Any], repeats: int) -> tuple[float, Any]:
    """Return median runtime and the final result from repeated calls."""
    values = []
    result = None
    for _ in range(repeats):
        start = time.perf_counter()
        result = fn()
        values.append(time.perf_counter() - start)
    return statistics.median(values), result


def _default_atom_type(atom_type_mask: int) -> int:
    if atom_type_mask != 0:
        return atom_type_mask
    return int(oegraphsim.OEFPAtomType_DefaultCircularAtom)


def _default_bond_type(bond_type_mask: int) -> int:
    if bond_type_mask != 0:
        return bond_type_mask
    return int(oegraphsim.OEFPBondType_DefaultCircularBond)


def build_oefp_batch(
    mols: Sequence[Any],
    *,
    numbits: int,
    min_distance: int,
    max_distance: int,
    atom_type_mask: int,
    bond_type_mask: int,
) -> Any:
    """Generate OEFP fingerprints with the same settings oecluster uses."""
    import oefp

    atom_type = _default_atom_type(atom_type_mask)
    bond_type = _default_bond_type(bond_type_mask)
    fps = []
    for mol in mols:
        oe_fp = oegraphsim.OEFingerPrint()
        if not oegraphsim.OEMakeCircularFP(
            oe_fp,
            mol,
            numbits,
            min_distance,
            max_distance,
            atom_type,
            bond_type,
        ):
            raise RuntimeError("OEMakeCircularFP failed")
        fps.append(oefp.from_openeye_fingerprint(oe_fp))
    return oefp.OEFPBatch.from_fingerprints(fps)


def _parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument("--count", type=int, default=512)
    parser.add_argument("--threads", type=int, default=0)
    parser.add_argument("--chunk-size", type=int, default=256)
    parser.add_argument("--repeats", type=int, default=3)
    parser.add_argument("--numbits", type=int, default=2048)
    parser.add_argument("--min-distance", type=int, default=0)
    parser.add_argument("--max-distance", type=int, default=2)
    parser.add_argument("--atom-type-mask", type=int, default=0)
    parser.add_argument("--bond-type-mask", type=int, default=0)
    parser.add_argument("--max-ratio", type=float, default=1.05)
    parser.add_argument(
        "--oecluster-source-dir",
        type=Path,
        default=_DEFAULT_OECLUSTER_SOURCE_DIR,
    )
    return parser


def main() -> int:
    """Run the benchmark and fail if OEFP misses the configured ratio gate."""
    args = _parser().parse_args()
    _prepend_source_paths(args.oecluster_source_dir)

    import oecluster
    import oefp

    mols = build_molecules(args.count)
    batch = build_oefp_batch(
        mols,
        numbits=args.numbits,
        min_distance=args.min_distance,
        max_distance=args.max_distance,
        atom_type_mask=args.atom_type_mask,
        bond_type_mask=args.bond_type_mask,
    )
    metric = oefp.Metric.jaccard()

    oe_time, oe_dist = time_call(
        lambda: oecluster.pdist(
            mols,
            "fingerprint",
            similarity=False,
            num_threads=args.threads,
            chunk_size=args.chunk_size,
            fp_type="circular",
            numbits=args.numbits,
            min_distance=args.min_distance,
            max_distance=args.max_distance,
            atom_type_mask=args.atom_type_mask,
            bond_type_mask=args.bond_type_mask,
            similarity_func="tanimoto",
        ),
        args.repeats,
    )
    fp_time, fp_dist = time_call(
        lambda: oefp.pdist(
            batch,
            metric,
            num_threads=args.threads,
            chunk_size=args.chunk_size,
        ),
        args.repeats,
    )

    oe_values = np.asarray(oe_dist)
    fp_values = np.asarray(fp_dist)
    np.testing.assert_allclose(fp_values, oe_values, rtol=1e-7, atol=1e-7)

    ratio = fp_time / oe_time if oe_time > 0 else float("inf")
    print(f"oecluster pdist median: {oe_time:.6f}s")
    print(f"oefp pdist median:      {fp_time:.6f}s")
    print(f"oefp/oecluster ratio:   {ratio:.3f}")

    if ratio > args.max_ratio:
        raise SystemExit(
            f"OEFP pdist ratio {ratio:.3f} exceeds configured gate {args.max_ratio:.3f}"
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/benchmark_oecluster_fingerprint.cpp
#include "oefp/batch.h"
#include "oefp/compare.h"
#include "oefp/metric.h"
#include "oefp/openeye.h"

#include "oecluster/CDist.h"
#include "oecluster/PDist.h"
#include "oecluster/StorageBackend.h"
#include "oecluster/metrics/FingerprintMetric.h"

#include <oechem.h>
#include <oegraphsim.h>

#include <chrono>
#include <cmath>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

std::vector<std::string> seed_smiles() {
    return {
        "c1ccccc1",
        "c1ccc(O)cc1",
        "c1ccncc1",
        "CCCCCCCC",
        "CCO",
        "CCN",
        "CC(=O)O",
        "c1ccc(Cl)cc1",
    };
}

std::vector<OEChem::OEGraphMol> build_molecules(std::size_t count) {
    const auto seeds = seed_smiles();
    std::vector<OEChem::OEGraphMol> mols;
    mols.reserve(count);
    for (std::size_t index = 0; index < count; ++index) {
        OEChem::OEGraphMol mol;
        if (!OEChem::OESmilesToMol(mol, seeds[index % seeds.size()])) {
            throw std::runtime_error("OESmilesToMol failed.");
        }
        mols.push_back(mol);
    }
    return mols;
}

std::vector<OEChem::OEMolBase*> mol_ptrs(std::vector<OEChem::OEGraphMol>& mols) {
    std::vector<OEChem::OEMolBase*> ptrs;
    ptrs.reserve(mols.size());
    for (auto& mol : mols) {
        ptrs.push_back(&static_cast<OEChem::OEMolBase&>(mol));
    }
    return ptrs;
}

OEFP::OEFPBatch build_oefp_batch(const std::vector<OEChem::OEGraphMol>& mols) {
    std::vector<OEFP::OEFP> fps;
    fps.reserve(mols.size());
    for (const auto& mol : mols) {
        OEGraphSim::OEFingerPrint oe_fp;
        if (!OEGraphSim::OEMakeCircularFP(
                oe_fp,
                mol,
                2048u,
                0u,
                2u,
                OEGraphSim::OEFPAtomType::DefaultCircularAtom,
                OEGraphSim::OEFPBondType::DefaultCircularBond)) {
            throw std::runtime_error("OEMakeCircularFP failed.");
        }
        fps.push_back(OEFP::FromOEFingerPrint(oe_fp));
    }
    return OEFP::OEFPBatch::FromFingerprints(fps);
}

template <typename Func>
double elapsed_seconds(Func&& func) {
    const auto start = std::chrono::steady_clock::now();
    func();
    const auto stop = std::chrono::steady_clock::now();
    return std::chrono::duration<double>(stop - start).count();
}

void assert_close(const std::vector<double>& a, const double* b, std::size_t size) {
    if (a.size() != size) {
        throw std::runtime_error("OEFP and oecluster result sizes differ.");
    }
    for (std::size_t index = 0; index < size; ++index) {
        if (std::fabs(a[index] - b[index]) > 1e-7) {
            throw std::runtime_error("OEFP and oecluster results differ.");
        }
    }
}

void assert_close(const std::vector<double>& a, const std::vector<double>& b) {
    if (a.size() != b.size()) {
        throw std::runtime_error("OEFP and oecluster result sizes differ.");
    }
    for (std::size_t index = 0; index < a.size(); ++index) {
        if (std::fabs(a[index] - b[index]) > 1e-7) {
            throw std::runtime_error("OEFP and oecluster results differ.");
        }
    }
}

} // namespace

int main(int argc, char** argv) {
    const auto count = argc > 1 ? static_cast<std::size_t>(std::stoul(argv[1])) : 512u;
    const auto threads = argc > 2 ? static_cast<std::size_t>(std::stoul(argv[2])) : 0u;
    const auto chunk_size = argc > 3 ? static_cast<std::size_t>(std::stoul(argv[3])) : 256u;

    auto mols = build_molecules(count);
    auto ptrs = mol_ptrs(mols);
    const auto batch = build_oefp_batch(mols);
    const auto metric = OEFP::Metric::Jaccard();
    OEFP::BatchKernelOptions oe_options;
    oe_options.num_threads = threads;
    oe_options.chunk_size = chunk_size;

    OECluster::FingerprintOptions fp_options;
    fp_options.fp_type = "circular";
    fp_options.numbits = 2048u;
    fp_options.min_distance = 0u;
    fp_options.max_distance = 2u;
    fp_options.similarity_func = "tanimoto";
    fp_options.similarity = false;

    OECluster::PDistOptions pd_options;
    pd_options.num_threads = threads;
    pd_options.chunk_size = chunk_size;

    OECluster::FingerprintMetric pd_metric(ptrs, fp_options);
    OECluster::DenseStorage storage(count);

    const double oe_pdist_time = elapsed_seconds([&]() {
        OECluster::pdist(pd_metric, storage, pd_options);
    });
    std::vector<double> oe_fp_pdist;
    const double fp_pdist_time = elapsed_seconds([&]() {
        oe_fp_pdist = OEFP::PDist(batch, metric, oe_options);
    });
    assert_close(oe_fp_pdist, storage.Data(), storage.NumPairs());

    const std::size_t n_a = count / 2u;
    std::vector<OEChem::OEGraphMol> mols_a(
        mols.begin(),
        mols.begin() + static_cast<long>(n_a));
    std::vector<OEChem::OEGraphMol> mols_b(
        mols.begin() + static_cast<long>(n_a),
        mols.end());
    const auto batch_a = build_oefp_batch(mols_a);
    const auto batch_b = build_oefp_batch(mols_b);
    std::vector<OEChem::OEMolBase*> cd_ptrs;
    cd_ptrs.reserve(mols.size());
    for (auto& mol : mols_a) {
        cd_ptrs.push_back(&static_cast<OEChem::OEMolBase&>(mol));
    }
    for (auto& mol : mols_b) {
        cd_ptrs.push_back(&static_cast<OEChem::OEMolBase&>(mol));
    }

    OECluster::FingerprintMetric cd_metric(cd_ptrs, fp_options);
    OECluster::CDistOptions cd_options;
    cd_options.num_threads = threads;
    cd_options.chunk_size = chunk_size;
    std::vector<double> oe_cd_out(n_a * (count - n_a), 0.0);

    const double oe_cdist_time = elapsed_seconds([&]() {
        OECluster::cdist(cd_metric, n_a, oe_cd_out.data(), cd_options);
    });
    std::vector<double> oe_fp_cd;
    const double fp_cdist_time = elapsed_seconds([&]() {
        oe_fp_cd = OEFP::CDist(batch_a, batch_b, metric, oe_options);
    });
    assert_close(oe_fp_cd, oe_cd_out);

    std::cout << "oecluster pdist: " << oe_pdist_time << "s\n";
    std::cout << "oefp pdist:      " << fp_pdist_time << "s\n";
    std::cout << "oecluster cdist: " << oe_cdist_time << "s\n";
    std::cout << "oefp cdist:      " << fp_cdist_time << "s\n";

    if (fp_pdist_time > oe_pdist_time * 1.05) {
        throw std::runtime_error("OEFP pdist is more than 5% slower than oecluster.");
    }
    if (fp_cdist_time > oe_cdist_time * 1.05) {
        throw std::runtime_error("OEFP cdist is more than 5% slower than oecluster.");
    }
    return 0;
}
